From: x0158321 <ashuba@uglx0158321.(none)>
Date: Mon, 6 Jun 2011 14:10:56 +0000 (+0300)
Subject: OMAP4: DSS HDMI: HDMI cable unplug issue on video playback
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=73e9c211e4b1f7e33e83f8baf72f2e99717a86ad

OMAP4: DSS HDMI: HDMI cable unplug issue on video playback

Patch will fix HDMI cable unplug issue during video playback.
Issue looks like Ducati crash.

Ducati crash during video playback might be caused by two reasons below:
1) Too late call to the hdmi_lib_stop_acr_wa()
2) Receiving by Ducati two HDMI ACR WA notify events with payload=0.
This patch will fix both.

Change-Id: Ibcc8ca0b1d7cf3c851627af31a0c89b1290c56e7
Signed-off-by: Anatolii Shuba <x0158321@ti.com>
Signed-off-by: Taras Kondratiuk <x0153363@ti.com>
---

Index: kernel/arch/arm/plat-omap/hdmi_lib.c
===================================================================
--- kernel.orig/arch/arm/plat-omap/hdmi_lib.c	2011-07-01 18:10:58.783401747 +0530
+++ kernel/arch/arm/plat-omap/hdmi_lib.c	2011-07-01 18:11:01.669648880 +0530
@@ -232,6 +232,7 @@
 	struct omap_chip_id audio_wa_chip_ids;
 	struct task_struct *wa_task;
 	u32 ack_payload;
+	bool audio_wa_started; /* HDMI WA guard*/
 #endif
 	u32 pixel_clock;
 } hdmi;
@@ -1457,11 +1458,23 @@
 }
 int hdmi_lib_start_acr_wa(void)
 {
-	return hdmi_lib_acr_wa_send_event(hdmi.cts_interval);
+	int ret = 0;
+	if (!hdmi.audio_wa_started) {
+		ret = hdmi_lib_acr_wa_send_event(hdmi.cts_interval);
+		if (!ret)
+			hdmi.audio_wa_started = true;
+	}
+	return ret;
 }
 int hdmi_lib_stop_acr_wa(void)
 {
-	return hdmi_lib_acr_wa_send_event(0);
+	int ret = 0;
+	if (hdmi.audio_wa_started) {
+		ret = hdmi_lib_acr_wa_send_event(0);
+		if (!ret)
+			hdmi.audio_wa_started = false;
+	}
+	return ret;
 }
 
 void hdmi_notify_event_ack_func(u16 proc_id, u16 line_id, u32 event_id,
@@ -1778,6 +1791,7 @@
 	hdmi.notify_event_reg = HDMI_NOTIFY_EVENT_NOTREG;
 	hdmi.audio_wa_chip_ids.oc = CHIP_IS_OMAP4430ES2 |
 			CHIP_IS_OMAP4430ES2_1 | CHIP_IS_OMAP4430ES2_2;
+	hdmi.audio_wa_started = false;
 #endif
 
 //	mutex_init(&hdmi.mutex);
@@ -2062,6 +2076,7 @@
 	hdmi_w1_audio_stop();
 	/* if audio is not used, switch to smart-idle & wakeup capable*/
 	REG_FLD_MOD(HDMI_WP, HDMI_WP_SYSCONFIG, 0x3, 3, 2);
+	printk(KERN_INFO "Stop audio transfer...\n");
 	return 0;
 }
 
Index: kernel/drivers/video/omap2/dss/hdmi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/hdmi.c	2011-07-01 18:11:00.809648873 +0530
+++ kernel/drivers/video/omap2/dss/hdmi.c	2011-07-01 18:11:01.669648880 +0530
@@ -1439,6 +1439,9 @@
 		goto retry;
 	}
 
+#ifdef CONFIG_OMAP_HDMI_AUDIO_WA
+	hdmi_lib_stop_acr_wa();
+#endif
 	return 0;
 }
 
